src/approaches/grammar_search_invention_approach.py

```diff
@@ -49,44 +50,109 @@ def _generate(self) -> Iterator[Tuple[Predicate, float]]:
 
 
 @dataclass(frozen=True, eq=False, repr=False)
-class _SingleAttributeGEClassifier:
-    """Check whether a single attribute value on an object is >= some value.
+class _SingleAttributeCompareClassifier:
+    """Compare a single feature value with a constant value.
     """
     object_index: int
     attribute_name: str
-    value: float
+    constant: float
+    compare: Callable[[float, float], bool]
+    compare_str: str
 
     def __call__(self, s: State, o: Sequence[Object]) -> bool:
         obj = o[self.object_index]
-        return s.get(obj, self.attribute_name) >= self.value
+        return self.compare(s.get(obj, self.attribute_name), self.constant)
+
+    def __str__(self) -> str:
+        return f"({self.object_index}.{self.attribute_name}" + \
+               f"{self.compare_str}{self.constant:.3})"
 
 
 @dataclass(frozen=True, eq=False, repr=False)
 class _HoldingDummyPredicateGrammar(_PredicateGrammar):
     """A hardcoded cover-specific grammar.
+
     Good for testing with:
         python src/main.py --env cover --approach grammar_search_invention \
             --seed 0 --excluded_predicates Holding
     """
     def _generate(self) -> Iterator[Tuple[Predicate, float]]:
         # A necessary predicate
-        name = "InventedHolding"
         block_type = [t for t in self.types if t.name == "block"][0]
         types = [block_type]
-        classifier = _SingleAttributeGEClassifier(0, "grasp", -0.9)
-        yield (Predicate(name, types, classifier), 1.)
+        classifier = _SingleAttributeCompareClassifier(0, "grasp", -0.9,
+                                                       ge, ">=")
+        yield (Predicate(str(classifier), types, classifier), 1.)
 
         # An unnecessary predicate (because it's redundant)
-        name = "InventedDummy"
-        block_type = [t for t in self.types if t.name == "block"][0]
-        types = [block_type]
-        classifier = _SingleAttributeGEClassifier(0, "is_block", 0.5)
-        yield (Predicate(name, types, classifier), 1.)
+        classifier = _SingleAttributeCompareClassifier(0, "is_block", 0.5,
+                                                       ge, ">=")
+        yield (Predicate(str(classifier), types, classifier), 1.)
+
+
+def _halving_constant_generator(lo: float, hi: float) -> Iterator[float]:
+    mid = (hi + lo) / 2.
+    yield mid
+    left_gen = _halving_constant_generator(lo, mid)
+    right_gen = _halving_constant_generator(mid, hi)
+    for l, r in zip(left_gen, right_gen):
+        yield l
+        yield r
```

Comment:
instead of halving_constant_generator, is there any reason you didn't just do np.linspace? is it because you want to prioritize splitting at half points, then quarter points, etc? if so, then why are 0 and 1 the first elements, shouldn't we start with 0.5?

Reply:
yeah that's why. and yeah I guess starting with 0.5 would make sense, but I couldn't think of a way to easily code that. since this is all likely temporary gonna leave it for now